### PR TITLE
Escape userid

### PR DESCRIPTION
**Pull request by Half-Shot**
...if the user contains characters not supported by matrix.

### CODE REVIEW

File: lib/models/users/matrix.js [hunk 1]
```diff
@@ -69,6 +75,23 @@ MatrixUser.prototype.serialize = function() {
...
+/**
+ * Make a userId conform to the matrix spec using QP escaping.
+ * Grammar taken from: https://matrix.org/docs/spec/appendices.html#identifier-grammar
+ */
+MatrixUser.prototype.escapeUserId = function() {
+    // Currently Matrix accepts / in the userId, although going forward it will be removed.
+    const badChars = new Set(this.localpart.replace(/([A-z0-9]|-|\.|=|_)+/g, ""));
+    let res = this.localpart;
+    badChars.forEach((c) => {
+        const hex = c.charCodeAt(0).toString(16).toUpperCase();
```
Comment (turt2live) [inline lib/models/users/matrix.js]: wouldn't we want to lowercase the hex?

File: spec/unit/matrix-user.spec.js [hunk 2]
```diff
@@ -0,0 +1,35 @@
+const MatrixUser = require("../../lib/models/users/matrix");
+
+describe("MatrixUser", function() {
+    describe("escapeUserId", function() {
+        it("should not escape acceptable userids", function() {
```
Comment (turt2live) [inline spec/unit/matrix-user.spec.js]: Please keep `user` and `ids` as two words, preferably also "IDs". Same applies to below.

File: spec/unit/matrix-user.spec.js [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+            [new MatrixUser("@test:localhost", null, false),
+            new MatrixUser("@42:localhost", null, false),
```
Comment (turt2live) [inline spec/unit/matrix-user.spec.js]: indentation.

File: spec/unit/matrix-user.spec.js [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+        it("should escape unacceptable userids", function() {
+            const expected = [
+                "@=24:localhost",
+                "@500=24=20dog:localhost",
+                "@woah=2A=2A=2A:localhost",
+                "@=D83D:localhost",
+                "@matrix.org=2Fspec:localhost",
+            ];
+            [new MatrixUser("@$:localhost", null, false),
+            new MatrixUser("@500$ dog:localhost", null, false),
```
Comment (turt2live) [inline spec/unit/matrix-user.spec.js]: indentation

**Comment by turt2live** [comment]
@Half-Shot ping

File: lib/models/users/matrix.js (hunk 1 above)
Comment (Half-Shot) [reply lib/models/users/matrix.js]: I prefer uppercase hex, but I don't know if we have a standard for it.

**Pull request merged**